File: nvim/oxi/core/src/key_map.rs
use nvim_oxi::{
    api::{self, opts::SetKeymapOpts, types::Mode},
    Error,
};

struct Keymap {
    mode: Mode,
    lhs: String,
    rhs: String,
    opts: SetKeymapOpts,
}

impl Keymap {
    fn new(mode: Mode, lhs: impl Into<String>, rhs: impl Into<String>) -> Keymap {
        let opts = SetKeymapOpts::builder().noremap(true).silent(true).build();
        Keymap {
            mode,
            lhs: lhs.into(),
            rhs: rhs.into(),
            opts,
        }
    }
}

pub fn init() -> Result<(), Error> {
    let maps = [
        Keymap::new(Mode::Normal, "<space>", ""),
        Keymap::new(Mode::Visual, "<space>", ""),
        Keymap::new(Mode::Normal, "<space>w", "<cmd>silent w!<cr>"),
        Keymap::new(Mode::Normal, "<space>q", "<cmd>q!<cr>"),
        Keymap::new(Mode::Normal, "<space>wq", "<cmd>wq<cr>"),
        Keymap::new(Mode::Normal, r"\", ","),
        Keymap::new(Mode::Normal, "j", "gj"),
        Keymap::new(Mode::Normal, "k", "gk"),
        Keymap::new(Mode::Normal, "gj", "j"),
        Keymap::new(Mode::Normal, "gk", "k"),
        Keymap::new(Mode::Normal, "ZZ", ""),
        Keymap::new(Mode::Normal, "ZQ", ""),
        Keymap::new(Mode::Normal, "<space>tt", "<cmd>terminal<CR>"),
        Keymap::new(Mode::Normal, "<space>si", r#"<cmd>%s/"/'/g<CR>"#),
        Keymap::new(Mode::Insert, "jj", "<esc>"),
        Keymap::new(Mode::Insert, "<C-l>", "<esc>"),
        Keymap::new(Mode::InsertCmdLine, "<C-l>", "<esc>"),
        Keymap::new(Mode::Terminal, "<M-i>", "<C-\\><C-n>"),
        Keymap::new(
            Mode::Normal,
            "<leader>fp",
            r#"vim.api.nvim_command('let @" = expand("%:.")')"#,
        ),
        Keymap::new(
            Mode::Normal,
            "<leader>fd",
            r#"vim.api.nvim_command('let @" = expand("%:p:h")')"#,
        ),
        Keymap::new(
            Mode::Normal,
            "<leader>fe",
            r#"vim.api.nvim_command('echo expand("%:p")')"#,
        ),
        Keymap::new(
            Mode::Normal,
            "<leader>nn",
            r#"vim.api.nvim_command('set number!')"#,
        ),
        Keymap::new(Mode::Normal, "<leader>hl", "<cmd>nohl<cr>"),
        Keymap::new(Mode::Normal, "<leader>o", "<cmd>only<cr>"),
    ];

    for map in maps {
        api::set_keymap(map.mode, &map.lhs, &map.rhs, &map.opts)?;
    }

    Ok(())
}

File: nvim/lua/core/keymap.lua
local utils = require('utils')

local maps = {
  { { 'n', 'v' }, '<space>', '' },
  { 'n', '<space>w', '<cmd>silent w!<cr>' },
  { 'n', '<space>q', '<cmd>q!<cr>' },
  { 'n', '<space>wq', '<cmd>wq<cr>' },
  { 'n', [[\]], ',' },
  { 'n', 'j', 'gj' },
  { 'n', 'k', 'gk' },
  { 'n', 'gj', 'j' },
  { 'n', 'gk', 'k' },
  { 'n', 'ZZ', '' },
  { 'n', 'ZQ', '' },
  { 'n', '<space>tt', '<cmd>terminal<CR>' },
  { 'n', '<space>si', [[<cmd>%s/"/'/g<CR>]] },
  { 'i', 'jj', '<esc>' },
  { { 'i', 'c' }, '<C-l>', '<esc>' },
  { 't', '<M-i>', [[<C-\><C-n>]] },
  { 'n', '<space>j', '<cmd>bnext<cr>' },
  { 'n', '<space>k', '<cmd>bprevious<cr>' },
  {
    -- yank and copy current file's absolute path to clipboard
    'n',
    '<Leader>fp',
    function()
      vim.api.nvim_command('let @" = expand("%:.")')
      vim.api.nvim_command('let @+ = expand("%:.")')
    end,
  },
  {
    -- yank and copy directiory that current file existing to clipboard
    'n',
    '<Leader>fd',
    function()
      vim.api.nvim_command('let @" = expand("%:p:h")')
      vim.api.nvim_command('let @+ = expand("%:p:h")')
    end,
  },
  {
    -- echo current file name
    'n',
    '<Leader>fe',
    function()
      vim.api.nvim_command('echo expand("%:p")')
    end,
  },
  {
    -- toggle line number (call :set number!)
    'n',
    '<Leader>nn',
    function()
      vim.api.nvim_command('set number!')
    end,
  },
  {
    -- TODO: toggle hlsearch
    'n',
    '<Leader>hl',
    '<cmd>nohl<cr>',
  },
  {
    'n',
    '<Leader>o',
    '<cmd>only<cr>',
  },
}

utils.keymap_set(maps)

utils.create_custome_command('VV', 'tabnew ' .. utils.get_init_lua() .. ' | :tcd %:h', { nargs = 0 })
utils.create_custome_command('SV', 'source ' .. utils.get_init_lua(), { nargs = 0 })

utils.create_custome_command('SamBuild', '!sam build', { nargs = 0 })
